megamek/src/megamek/client/ratgenerator/ForceDescriptor.java

```diff
@@ -0,0 +1,1696 @@
+						if (mRec != null && weights.contains(mRec.getWeightClass())
+								&& RATGenerator.getInstance().findModelAvailabilityRecord(era, model, faction) != null) {
+							av = RATGenerator.getInstance().findChassisAvailabilityRecord(era, mRec.getChassisKey(), faction, getYear());
+							if (av == null) {
+								for (String alt : RATGenerator.getInstance().getFaction(faction).getParentFactions()) {
+									av = RATGenerator.getInstance().findChassisAvailabilityRecord(era, mRec.getChassisKey(), alt, getYear());
+									if (av != null) {
+										break;
+									}
+								}
+							}
+							if (Compute.d6(2) >= target - ((av == null)?0:av.adjustForRating(ratingLevel, totalLevels))) {
+								sub.setUnit(mRec);
+								if (useWeights) {
+									weights.remove((Object)mRec.getWeightClass());
+								}
+								foundUnit = true;
+								break;
+							}
+						}
+					}
+				}
+				if (!foundUnit && weights.contains(baseModel.getWeightClass())) {
+					av = RATGenerator.getInstance().findChassisAvailabilityRecord(era, baseModel.getChassisKey(), faction, getYear());
+					if (av == null) {
+						for (String alt : RATGenerator.getInstance().getFaction(faction).getParentFactions()) {
+							av = RATGenerator.getInstance().findChassisAvailabilityRecord(era, baseModel.getChassisKey(), alt, getYear());
+							if (av != null) {
+								break;
+							}
+						}
+					}
+					if (Compute.d6(2) >= target - ((av == null)?0:av.adjustForRating(ratingLevel, totalLevels))) {
+						sub.getChassis().add(baseModel.getChassis());
+						sub.setWeightClass(-1);
+						unit = sub.generate();
+						if (unit != null) {
+							sub.setUnit(unit);
+							if (useWeights) {
+								weights.remove((Object)sub.getWeightClass());
+							}
+							foundUnit = true;
+						}
+					} else if (ut == UnitType.TANK && Compute.d6(2) >= target - 6) {
+						if (useWeights) {
+							switch (baseModel.getMechSummary().getUnitSubType()) {
+							case "Hover":
+								if (weights.contains(EntityWeightClass.WEIGHT_HEAVY)) {
+									break;
+								}
+								/* fall through */
+							case "Wheeled":
+								if (weights.contains(EntityWeightClass.WEIGHT_ASSAULT)) {
+									break;
+								}
+								sub.getMovementModes().add(baseModel.getMovementMode());
+							}
+						}
+					} else if (ut == UnitType.INFANTRY) {
+						sub.getMovementModes().add(baseModel.getMovementMode());
+					}
+				}
+			}
+			if (!foundUnit) {
+				if (!weights.contains(sub.getWeightClass())) {
+					sub.setWeightClass(weights.get(0));
+				}
+				unit = sub.generate();
+				if (unit == null) {
+					sub.getMovementModes().clear();
+					unit = sub.generate();
+				}
+				if (unit != null) {
+					sub.setUnit(unit);
+					if (useWeights) {
+						weights.remove((Object)sub.getWeightClass());
+					}
+				}
+			}
+			if (ut == null || ut == UnitType.MEK) {
+				baseModel = null;
+			}
+		}
+	}
+	
+	public void setUnit(ModelRecord unit) {
```

Comment:
This method could probably use some comments. Also, the indentation seems off near the bottom.